backend/src/pilstark/estark.rs

```diff
@@ -0,0 +1,101 @@
+use crate::{pilstark, BackendImpl};
+use ast::analyzed::Analyzed;
+use number::{BigInt, DegreeType, FieldElement, GoldilocksField};
+
+use starky::{
+    merklehash::MerkleTreeGL,
+    polsarray::{PolKind, PolsArray},
+    stark_gen::StarkProof,
+    stark_setup::StarkSetup,
+    transcript::TranscriptGL,
+    types::{StarkStruct, Step, PIL},
+};
+
+pub struct EStark {
+    params: StarkStruct,
+}
+
+impl<F: FieldElement> BackendImpl<F> for EStark {
+    /// Creates our default configuration stark struct.
+    fn new(degree: DegreeType) -> Self {
+        if F::modulus().to_arbitrary_integer() != GoldilocksField::modulus().to_arbitrary_integer()
+        {
+            unimplemented!("eSTARK is only implemented for Goldilocks field");
+        }
+
+        assert_ne!(degree, 0);
+        let degree_bits = (DegreeType::BITS - (degree - 1).leading_zeros()) as usize;
+        let params = StarkStruct {
+            nBits: degree_bits,
+            nBitsExt: degree_bits + 1,
+            nQueries: 2,
+            verificationHashType: "GL".to_owned(),
+            steps: vec![Step { nBits: 19 }, Step { nBits: 17 }, Step { nBits: 7 }],
```

Comment:
@lvella here the first element in steps, namely 19, should be equal to nBitsExt.

Reply:
Could this eventually be enforced in the `StarkStruct` type?
```
fn steps(&self) -> Iterator<Item = Step> {
   once(Step::from(nBitsExt)).chain(self.steps)
}
```

Reply:
Yes. For the left elements in `steps` can be calculated by `steps[i+1] = steps[i] - by_step` with by_step customized, like 4.

Reply:
But what is the full rule set? I already know of this:
- Must be at least 2 steps;
- The first must be equal nBitsExt.

Can the number of bits be equal between steps, in case nBitsExt is already too small? What is the minimum `nBits`?

Reply:
the `nBits` should equal to the d strictly, where 2^d = N, N is the argument of main in PIL.   

For the by_step above can be defined as your well, the smaller, the safer, but slower.